kotlin/download-preplanned-map-area/src/main/java/com/esri/arcgisruntime/sample/downloadpreplannedmaparea/MainActivity.kt

```diff
@@ -0,0 +1,343 @@
+      val offlineMapParametersFuture =
+        offlineMapTask.createDefaultDownloadPreplannedOfflineMapParametersAsync(
+          mSelectedPreplannedMapArea
+        )
+      offlineMapParametersFuture?.addDoneListener {
+        try {
+          // get the offline map parameters
+          val offlineMapParameters = offlineMapParametersFuture.get()
+          // set the update mode to not receive updates
+          offlineMapParameters.updateMode = PreplannedUpdateMode.NO_UPDATES
+          // create a job to download the preplanned offline map to a temporary directory
+          downloadPreplannedOfflineMapJob = offlineMapTask.downloadPreplannedOfflineMap(
+            offlineMapParameters,
+            offlineMapDirectory.path + File.separator + mSelectedPreplannedMapArea?.portalItem?.title
+          ).also {
+            // create and update a progress dialog for the job
+            dialog = createProgressDialog(it)
+            dialog?.show()
+            it.addProgressChangedListener {
+              dialog?.progressBar?.progress = it.progress
+              dialog?.progressTextView?.text = "${it.progress}%"
+            }
+            // start the job
+            it.start()
+          }
+
+          // when the job finishes
+          downloadPreplannedOfflineMapJob?.addJobDoneListener {
+            // dismiss progress dialog
+            dialog?.dismiss()
+            // if there's a result from the download preplanned offline map job
+            if (downloadPreplannedOfflineMapJob?.status != Job.Status.SUCCEEDED) {
+              val error =
+                "Job finished with an error: " + downloadPreplannedOfflineMapJob?.error?.message
+              Toast.makeText(this, error, Toast.LENGTH_LONG).show()
+              Log.e(TAG, error)
+              return@addJobDoneListener
+            }
+            downloadPreplannedOfflineMapJob?.result?.let { downloadPreplannedOfflineMapResult ->
+              if (downloadPreplannedOfflineMapResult.hasErrors()) {
+                // collect the layer and table errors into a single alert message
+                val stringBuilder = StringBuilder("Errors: ")
+                downloadPreplannedOfflineMapResult.layerErrors?.forEach { (key, value) ->
+                  stringBuilder.append("Layer: ").append(key.name).append(". Exception: ")
+                    .append(value.message).append(". ")
+                }
+                downloadPreplannedOfflineMapResult.tableErrors?.forEach { (key, value) ->
+                  stringBuilder.append("Table: ").append(key.tableName).append(". Exception: ")
+                    .append(value.message).append(". ")
+                }
+                val error =
+                  "One or more errors occurred with the Offline Map Result: $stringBuilder"
+                Toast.makeText(this, error, Toast.LENGTH_LONG).show()
+                Log.e(TAG, error)
+                return@addJobDoneListener
```

Comment:
If it's necessary I understand, but all of the append statements make this quite difficult to understand. I suggest this as a simplification, but I'm not 100% on how stringbuilder operates.

```suggestion
                val stringBuilder = StringBuilder("Errors: ")
                downloadPreplannedOfflineMapResult.layerErrors?.forEach { (key, value) ->
                  stringBuilder.append("Layer: ${key.name}. Exception: ${value.message}. ")
                }
                downloadPreplannedOfflineMapResult.tableErrors?.forEach { (key, value) ->
                  stringBuilder.append("Table: ${key.tableName}. Exception: ${value.message}. ")
                }
                val error =
                  "One or more errors occurred with the Offline Map Result: $stringBuilder"
                Toast.makeText(this, error, Toast.LENGTH_LONG).show()
                Log.e(TAG, error)
                return@addJobDoneListener
```